playerctl/playerctl-cli.c

```diff
@@ -16,279 +16,359 @@
-
-    g_print("%s", value);
-
-    g_free(value);
-  } else if (g_strcmp0(command[0], "status") == 0) {
-    /* STATUS */
-    gchar *status = NULL;
-    g_object_get(player, "status", &status, NULL);
-
-    if (status) {
-      g_print("%s\n", status);
+  } else {
+    g_object_get(player, "position", &offset, NULL);
+    g_print("%g\n", (double)offset / 1000000);
+  }
+
+  return TRUE;
+}
+
+static gboolean set_or_get_volume (const gchar *value, PlayerctlPlayer *player, GError **error)
+{
+  const gchar *volume = value;
+  gdouble level;
+
+  if (volume) {
+    size_t last = strlen(volume) - 1;
+
+    if(volume[last] == '+' || volume[last] == '-') {
+      gdouble adjustment = g_ascii_strtod(volume, NULL);
+
+      if(volume[last] == '-') {
+          adjustment *= -1;
+      }
+
+      g_object_get(player, "volume", &level, NULL);
+      level += adjustment;
     } else {
-      g_print("Not available\n");
+      level = g_ascii_strtod(volume, NULL);
     }
-
-    g_free(status);
+    g_object_set(player, "volume", level, NULL);
   } else {
-    /* unrecognized command */
-    g_print("%s", g_option_context_get_help(context, TRUE, NULL));
+    g_object_get(player, "volume", &level, NULL);
+    g_print("%g\n", level);
   }
 
-  if (error != NULL) {
-    g_printerr("An error occurred: %s\n", error->message);
-    g_clear_error(&error);
-    g_object_unref(player);
-    return 1;
-  }
+  return TRUE;
+}
 
-  g_object_unref(player);
-  return 0;
+static gboolean status (const gchar *value, PlayerctlPlayer *player, GError **error)
+{
+  gchar *state = NULL;
+
+  g_object_get(player, "status", &state, NULL);
+  g_print("%s\n", state ? state : "Not available");
+  g_free(state);
+
+  return TRUE;
 }
 
-static gchar *list_player_names(GError **err)
+static gboolean get_metadata (const gchar *value, PlayerctlPlayer *player, GError **error)
 {
   GError *tmp_error = NULL;
+  gchar *data;
+
+  if (g_strcmp0(value, "artist") == 0)
+    data = playerctl_player_get_artist(player, &tmp_error);
+  else if (g_strcmp0(value, "title") == 0)
+    data = playerctl_player_get_title(player, &tmp_error);
+  else if (g_strcmp0(value, "album") == 0)
+    data = playerctl_player_get_album(player, &tmp_error);
+  else
+    data = playerctl_player_print_metadata_prop(player, value, &tmp_error);
+
+  if (tmp_error) {
+    g_propagate_error(error, tmp_error);
+    return FALSE;
+  }
 
-  GDBusProxy *proxy = g_dbus_proxy_new_for_bus_sync(
-      G_BUS_TYPE_SESSION,
-      G_DBUS_PROXY_FLAGS_NONE,
-      NULL,
-      "org.freedesktop.DBus",
-      "/org/freedesktop/DBus",
-      "org.freedesktop.DBus",
-      NULL,
-      &tmp_error);
+  g_print("%s", data);
+  g_free(data);
 
-  if (tmp_error != NULL) {
-    g_propagate_error(err, tmp_error);
-    return NULL;
+  return TRUE;
+}
+
+struct PlayerCommand {
+  const gchar *name;
+  gboolean (*func) (const gchar *value, PlayerctlPlayer *player, GError **error);
```

Comment:
This function signature is not extensible. It will not handle the case of nested subcommands or a subcommand with its own arguments. It should rather take a list of the remaining command line arguments.